graph-gateway/src/client_query.rs

```diff
@@ -95,57 +89,74 @@ pub async fn handle_query(
         }
-        QuerySelector::Deployment(_) => {
+        QuerySelector::Deployment(id) => {
             // Authorization is based on the "authorized subgraphs" allowlist. We need to resolve
             // the subgraph deployments to check if any of the deployment's subgraphs are
             // authorized, otherwise return an error.
-            let (deployments, subgraph) = resolve_subgraph_deployments(&ctx.network, &selector)?;
+            let info = ctx
+                .network
+                .resolve_with_deployment_id(id)
+                .map_err(Error::Internal)?
+                .ok_or_else(|| Error::SubgraphNotFound(anyhow!("deployment_id: {id}")))?;
 
             // If none of the deployment's subgraphs are authorized, return an error.
-            let deployment_subgraphs = deployments
-                .iter()
-                .flat_map(|d| d.subgraphs.iter())
-                .collect::<Vec<_>>();
-            if !auth.is_any_deployment_subgraph_authorized(&deployment_subgraphs) {
+            if !auth.is_any_deployment_subgraph_authorized(&info.subgraphs) {
                 return Err(Error::Auth(anyhow!("Deployment not authorized by user")));
             }
 
-            (deployments, subgraph)
+            info
         }
     };
 
-    tracing::info!(deployments = ?deployments.iter().map(|d| d.id).collect::<Vec<_>>());
-
-    if let Some(l2_url) = ctx.l2_gateway.as_ref() {
-        // Forward query to L2 gateway if it's marked as transferred & there are no allocations.
-        // abf62a6d-c071-4507-b528-ddc8e250127a
-        let transferred_to_l2 = deployments.iter().all(|d| d.transferred_to_l2);
-        if transferred_to_l2 {
+    tracing::info!(deployments = ?info.deployments);
+
+    if let Some(l2_gateway_url) = ctx.l2_gateway.as_ref() {
+        // Forward the query to L2 gateway if it's marked as transferred (and have no allocations).
+        //
+        // For a request with a subgraph ID selector, the resolved subgraph's associated
+        // deployments:
+        //  - All of them MUST be marked as transferred to L2.
+        //  - All of them MUST have NO allocations.
+        // For a request with a deployment ID selector, the resolved deployment:
+        //  - It MUST be marked as transferred to L2.
+        //  - It MUST have no allocations.
+        //
+        // In both cases, as deployments with no allocations are filtered-out by the network
+        // service, this is guaranteed. See ref: 43be510e-7157-43d6-b2cf-546abf400edf
+        if info.transferred_to_l2 {
             return Ok(forward_request_to_l2(
                 &ctx.indexer_client.client,
-                l2_url,
+                l2_gateway_url,
                 &original_uri,
                 headers,
                 payload,
-                subgraph.and_then(|s| s.l2_id),
+                info.id_on_l2,
             )
             .await);
         }
     }
```

Comment:
@Theodus The new network resolution implementation filters out deployments with zero allocations, meaning there are no valid indexers/candidates to respond to the requests. 

Under which circumstances is the `transferred_to_l2` field set to `true`? Is the following assumption correct?

> If allocations are associated with the deployment, the field remains set to `false`. When the number of associated allocations is 0, the transfer tools mark the deployment as transferred.

Reply:
I am trying to find a way to keep things simple without requiring any modifications to the new network topology service fetch logic 🤔

Reply:
Does this also imply that a user will receive a `SubgraphNotFound` error when there are no active allocations instead of a `NoIndexers` error? That seems wrong to me.

Reply:
It seems like `DeploymentInfo::allocations : Vec1<AllocationInfo>` is just an incorrect requirement. It makes sense for the Vec to potentially be empty.

Reply:
I agree 👍 

I am changing this in a way that we can easily remove the code once the L2 gateway forwarding is deprecated.